File: internal/state/state.go
package state

import (
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/minkuik/agentgit/internal/git"
)

func Dir(repoRoot string) (string, error) {
	gitDir, err := git.GitDir(repoRoot)
	if err != nil {
		return "", err
	}
	path := filepath.Join(gitDir, "agentgit")
	if err := os.MkdirAll(path, 0o755); err != nil {
		return "", err
	}
	return path, nil
}

func ActiveRequestFile(repoRoot string) (string, error) {
	dir, err := Dir(repoRoot)
	if err != nil {
		return "", err
	}
	return filepath.Join(dir, "active-request"), nil
}

func SetActiveRequest(repoRoot string, requestID int64) error {
	path, err := ActiveRequestFile(repoRoot)
	if err != nil {
		return err
	}
	return os.WriteFile(path, []byte(strconv.FormatInt(requestID, 10)+"\n"), 0o644)
}

func GetActiveRequest(repoRoot string) (int64, bool, error) {
	path, err := ActiveRequestFile(repoRoot)
	if err != nil {
		return 0, false, err
	}
	raw, err := os.ReadFile(path)
	if os.IsNotExist(err) {
		return 0, false, nil
	}
	if err != nil {
		return 0, false, err
	}
	text := strings.TrimSpace(string(raw))
	if text == "" {
		return 0, false, nil
	}
	id, err := strconv.ParseInt(text, 10, 64)
	if err != nil {
		return 0, false, err
	}
	return id, true, nil
}

func ClearActiveRequest(repoRoot string) error {
	path, err := ActiveRequestFile(repoRoot)
	if err != nil {
		return err
	}
	err = os.Remove(path)
	if os.IsNotExist(err) {
		return nil
	}
	return err
}

File: src/agentgit/state.py
from __future__ import annotations

from pathlib import Path

from .git import git_dir


def state_dir(repo_root: Path) -> Path:
    path = git_dir(repo_root) / "agentgit"
    path.mkdir(parents=True, exist_ok=True)
    return path


def active_request_file(repo_root: Path) -> Path:
    return state_dir(repo_root) / "active-request"


def set_active_request(repo_root: Path, request_id: int) -> None:
    active_request_file(repo_root).write_text(f"{request_id}\n", encoding="utf-8")


def get_active_request(repo_root: Path) -> int | None:
    path = active_request_file(repo_root)
    if not path.exists():
        return None
    raw = path.read_text(encoding="utf-8").strip()
    return int(raw) if raw else None


def clear_active_request(repo_root: Path) -> None:
    path = active_request_file(repo_root)
    if path.exists():
        path.unlink()
